File: src/main/java/org/example/bankramenserver/domain/recurring/presentation/dto/response/RecurringPaymentDetailResponse.java
package org.example.bankramenserver.domain.recurring.presentation.dto.response;

import org.example.bankramenserver.domain.recurring.domain.RecurringPayment;
import org.example.bankramenserver.domain.recurring.domain.RecurringPaymentTransaction;

import java.time.LocalDate;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;

public record RecurringPaymentDetailResponse(
        UUID recurringPaymentId,
        String name,
        Long amount,
        String categoryName,
        RecurringPayment.Cycle cycle,
        LocalDate nextBillingDate,
        Integer billingDay,
        RecurringPayment.RegistrationType registrationType,
        boolean confirmed,
        boolean active,
        List<TransactionItem> transactions
) {
    public static RecurringPaymentDetailResponse from(RecurringPayment recurringPayment) {
        return new RecurringPaymentDetailResponse(
                recurringPayment.getId(),
                recurringPayment.getName(),
                recurringPayment.getAmount(),
                recurringPayment.getCategory() != null
                        ? recurringPayment.getCategory().getDisplayName()
                        : null,
                recurringPayment.getCycle(),
                recurringPayment.getNextBillingDate().toLocalDate(),
                recurringPayment.getBillingDay(),
                recurringPayment.getRegistrationType(),
                recurringPayment.isConfirmed(),
                recurringPayment.isActive(),
                recurringPayment.getTransactions()
                        .stream()
                        .sorted(
                                Comparator.comparing(
                                        item -> item.getTransaction().getTransactionDate()
                                )
                        )
                        .map(TransactionItem::from)
                        .toList()
        );
    }

    public record TransactionItem(
            UUID transactionId,
            LocalDate transactionDate,
            String description,
            Long amount,
            RecurringPaymentTransaction.MatchType matchType
    ) {
        public static TransactionItem from(RecurringPaymentTransaction item) {
            return new TransactionItem(
                    item.getTransaction().getId(),
                    item.getTransaction().getTransactionDate(),
                    item.getTransaction().getDescription(),
                    item.getTransaction().getAmount(),
                    item.getMatchType()
            );
        }
    }
}

File: src/main/kotlin/org/example/balogserver/domain/recurring/presentation/dto/response/RecurringPaymentDetailResponse.kt
package org.example.balogserver.domain.recurring.presentation.dto.response

import org.example.balogserver.domain.recurring.domain.RecurringPayment
import org.example.balogserver.domain.recurring.domain.RecurringPaymentTransaction
import java.time.LocalDate
import java.util.UUID

data class RecurringPaymentDetailResponse(
    val recurringPaymentId: UUID, val name: String, val amount: Long, val categoryName: String?, val cycle: RecurringPayment.Cycle,
    val nextBillingDate: LocalDate, val billingDay: Int, val registrationType: RecurringPayment.RegistrationType, val confirmed: Boolean,
    val active: Boolean, val transactions: List<TransactionItem>,
) {
    companion object {
        @JvmStatic
        fun from(recurringPayment: RecurringPayment) = RecurringPaymentDetailResponse(
            recurringPayment.id!!, recurringPayment.name, recurringPayment.amount!!, recurringPayment.category?.displayName,
            recurringPayment.cycle, recurringPayment.nextBillingDate.toLocalDate(), recurringPayment.billingDay,
            recurringPayment.registrationType, recurringPayment.confirmed, recurringPayment.active,
            recurringPayment.transactions.sortedBy { it.transaction.transactionDate }.map { TransactionItem.from(it) },
        )
    }
    data class TransactionItem(val transactionId: UUID, val transactionDate: LocalDate, val description: String?, val amount: Long, val matchType: RecurringPaymentTransaction.MatchType) {
        companion object {
            @JvmStatic fun from(item: RecurringPaymentTransaction) = TransactionItem(item.transaction.id!!, item.transaction.transactionDate, item.transaction.description!!, item.transaction.amount!!, item.matchType)
        }
    }
}
